### PR TITLE
Fix for ignoring name parameter for Percolator queries

### LINKED ISSUE #40405: Percolate query ignores "name" parameter
**Elasticsearch version** (`bin/elasticsearch --version`): `6.5.4`

**Plugins installed**: `[analysis-icu]`

**JVM version** (`java -version`): `openjdk version "11.0.1" 2018-10-16`

**OS version** (`uname -a` if on a Unix-like system): `Linux 7a321a3f5b0f 4.4.0-119-generic #143-Ubuntu SMP Mon Apr 2 16:08:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`

Deployed as a customized Docker image (following [the documented procedure](https://www.elastic.co/guide/en/elasticsearch/reference/6.5/docker.html#_c_customized_image)) with only `analysis-icu` plugin installed beyond the default.

**Description of the problem including expected versus actual behavior**:

According to [the documentation](https://www.elastic.co/guide/en/elasticsearch/reference/6.5/query-dsl-percolate-query.html#_parameters_10) (and [given example](https://www.elastic.co/guide/en/elasticsearch/reference/6.5/query-dsl-percolate-query.html#_specifying_multiple_percolate_queries)), providing the `name` parameter in percolate query should result in `_percolator_document_slot` field being suffixed with that value if multiple percolate queries are supplied. However, it's not the case and the field is suffixed with the `field` parameter value as if no `name` was provided. Also, the documentation is inconsistent in specifying if the parameter name should be `name` or `_name`.

**Steps to reproduce**:

Issue the percolate query request for two existing documents from `my_index` index (from Kibana Dev Tools):

```
GET watchers/_search
{
  "query": {
    "bool": {
      "should": [
        {
          "percolate": {
            "field": "watcher.query",
            "index": "my_index",
            "type": "_doc",
            "id": "5bf8b8aee6b3da2b8d9f5be8",
            "name": "name1"
          }
        },
        {
          "percolate": {
            "field": "watcher.query",
            "index": "my_index",
            "type": "_doc",
            "id": "5bdaf10f67c85d283c174ad9",
            "name": "name2"
          }
        }
      ]
    }
  }
}
```

Response:

```
{
  "took" : 17,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 1,
    "max_score" : 2.0,
    "hits" : [
      {
        "_index" : "watchers",
        "_type" : "_doc",
        "_id" : "5c988b8adc112234588a07b5",
        "_score" : 2.0,
        "_source" : {
          "watcher.query" : {
            "bool" : {
              "must" : {
                "match_all" : { }
              },
              "filter" : {
                "bool" : {
                  "must" : [
                    {
                      "geo_distance" : {
                        "distance" : "2000m",
                        "gps" : {
                          "lat" : 50.132179,
                          "lon" : 14.510528
                        }
                      }
                    }
                  ]
                }
              }
            }
          },
          "watcher.user_id" : "5b85318b036be43903b294de",
          "watcher.search_id" : "5c988b8adc112234588a07b5",
          "watcher.period" : "immediate"
        },
        "fields" : {
          "_percolator_document_slot_watcher.query" : [
            0
          ]
        }
      }
    ]
  }
}
```

The `fields` object contains `_percolator_document_slot_watcher.query`, however there should be keys like `_percolator_document_slot_name1` and `_percolator_document_slot_name2` present.

What actually works for my use case but is not documented (at least in the context of percolate query) is providing the `_name` parameter. This works just like any other [named queries](https://www.elastic.co/guide/en/elasticsearch/reference/6.5/search-request-named-queries-and-filters.html) resulting in `matched_queries` key in the output:

Request:

```
GET watchers/_search
{
  "query": {
    "bool": {
      "should": [
        {
          "percolate": {
            "field": "watcher.query",
            "index": "my_index",
            "type": "_doc",
            "id": "5bf8b8aee6b3da2b8d9f5be8",
            "_name": "name1"
          }
        },
        {
          "percolate": {
            "field": "watcher.query",
            "index": "my_index",
            "type": "_doc",
            "id": "5bdaf10f67c85d283c174ad9",
            "_name": "name2"
          }
        }
      ]
    }
  }
}
```

Response:

```
{
  "took" : 21,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 1,
    "max_score" : 2.0,
    "hits" : [
      {
        "_index" : "watchers",
        "_type" : "_doc",
        "_id" : "5c988b8adc112234588a07b5",
        "_score" : 2.0,
        "_source" : {
          "watcher.query" : {
            "bool" : {
              "must" : {
                "match_all" : { }
              },
              "filter" : {
                "bool" : {
                  "must" : [
                    {
                      "geo_distance" : {
                        "distance" : "2000m",
                        "gps" : {
                          "lat" : 50.132179,
                          "lon" : 14.510528
                        }
                      }
                    }
                  ]
                }
              }
            }
          },
          "watcher.user_id" : "5b85318b036be43903b294de",
          "watcher.search_id" : "5c988b8adc112234588a07b5",
          "watcher.period" : "immediate"
        },
        "fields" : {
          "_percolator_document_slot_watcher.query" : [
            0
          ]
        },
        "matched_queries" : [
          "name2",
          "name1"
        ]
      }
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by gurkankaymak**
fixes #40405

### CODE REVIEW

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java [hunk 1]
```diff
@@ -130,12 +130,12 @@
...
 
     /**
-     * @deprecated use {@link #PercolateQueryBuilder(String, BytesReference, XContentType)} with the document content
+     * @deprecated use {@link #PercolateQueryBuilder(String, String, BytesReference, XContentType)} with the document content
      * type to avoid autodetection.
      */
     @Deprecated
-    public PercolateQueryBuilder(String field, String documentType, BytesReference document) {
-        this(field, documentType, Collections.singletonList(document), XContentHelper.xContentType(document));
+    public PercolateQueryBuilder(String field, String documentType, String name, BytesReference document) {
```
Comment (jimczi) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: `name` is not a mandatory argument so I don't think it should be added in the ctr. Could we set it on the rewritten query instead if the value is not null ?

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java [hunk 2]
```diff
@@ -519,7 +521,7 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) {
             if (source == null) {
                 return this; // not executed yet
             } else {
-                return new PercolateQueryBuilder(field, documentType, Collections.singletonList(source),
+                return new PercolateQueryBuilder(field, documentType, name, Collections.singletonList(source),
```
Comment (jimczi) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: `name` can be null so I think we should do:
```
  PercolateQueryBuilder rewritten = new new PercolateQueryBuilder(field, documentType, name, Collections.singletonList(source), XContentHelper.xContentType(source));
  if (name != null) {
    rewritten.setName(name);
  }
  return rewritten;
```

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java [hunk 3]
```diff
@@ -555,7 +557,9 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) {
                 listener.onResponse(null);
             }, listener::onFailure));
         });
-        return new PercolateQueryBuilder(field, documentType, documentSupplier::get);
+
+
+        return new PercolateQueryBuilder(field, documentType, name, documentSupplier::get);
```
Comment (jimczi) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: same here

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @gurkankaymak , I left some comments. Can you also add a test to ensure that the fix is correct ?

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java (hunk 1 above)
Comment (gurkankaymak) [reply modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: updated as you suggested

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java [hunk 4]
```diff
@@ -555,7 +559,13 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) {
                 listener.onResponse(null);
             }, listener::onFailure));
         });
-        return new PercolateQueryBuilder(field, documentType, documentSupplier::get);
+
```
Comment (jimczi) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: extra line ?

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks, the change looks good but we need a test to ensure that the fix is correct. Can you add one ?

**Comment by gurkankaymak** [comment]
> Thanks, the change looks good but we need a test to ensure that the fix is correct. Can you add one ?

sure, I just left, I'll add a test when I get back on my computer

**Review by jimczi — COMMENTED** [review]
> sure, I just left, I'll add a test when I get back on my computer

Sorry I thought that you were asking for another review. No rush at all and thanks for working on this.

**Comment by gurkankaymak** [comment]
hi @jimczi i've returned from the military service :)
and added tests as we last talked, could you please review the changes

**Comment by javanna** [comment]
test this please

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java [hunk 5]
```diff
@@ -654,6 +663,10 @@ XContentType getXContentType() {
         return documentXContentType;
     }
 
+    public String getQueryName() {
```
Comment (jimczi) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: There is a getter in `AbstractQueryBuilder#queryName` so no need to add a new function.

**Review by jimczi — CHANGES_REQUESTED** [review]
Sorry for the late review. Thanks for adding the tests, I left one comment that needs a fix but the change looks good.

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java (hunk 5 above)
Comment (gurkankaymak) [reply modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: I noticed that getter but it returns the `protected String queryName` field in the `AbstractQueryBuilder`, i need to test for the `private String name` field in the `PercolateQueryBuilder`

Comment (gurkankaymak) [reply modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: hi @jimczi as i stated in the previous comment `AbstractQueryBuilder#queryName` is a getter of another field that's why i added the `getQueryName`

do you need me to update any other thing to move this pr forward?

Comment (jimczi) [reply modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: Right, sorry for the confusion. So let's change the name to `getName` (for consistency with `setName`) and add javadocs ?

Comment (gurkankaymak) [reply modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: @jimczi changing name to `getName` will overwrite the `AbstractQueryBuilder.getName` which returns another field (`PercolateQueryBuilder` => `public static final String NAME`)

Comment (jimczi) [reply modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java]: ok got it now, sorry for being slow and thanks for explaining.

**Review by jimczi — APPROVED** [review]
Thanks @gurkankaymak! Can you merge with master and we'll check that the tests pass in CI ?

**Comment by elasticcla** [comment]
Hi @gurkankaymak, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in your Git [commit](https://github.com/elastic/elasticsearch/pull/40843.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by gurkankaymak** [comment]
thanks for review @jimczi I synced branch with the master, can you start the CI tests

**Comment by jimczi** [comment]
> I synced branch with the master, can you start the CI tests
Sure I will but did you see https://github.com/elastic/elasticsearch/pull/40843#issuecomment-496104420 ? You need to add your e-mail into your github profile or use the same e-mail than the one that is already present.

**Comment by gurkankaymak** [comment]
the email is the same as before and exists in my profile, I pushed the sync code from different computer and i think i used the github username part instead of the full email, not sure how to fix this

**Comment by jimczi** [comment]
> , I pushed the sync code from different computer and i think i used the github username part instead of the full email, not sure how to fix this

Can you try to rebase to keep your first commit only (which contains the correct username) ?

**Comment by gurkankaymak** [comment]
@jimczi I'm new to rebase and i think i did not use it correctly, currently i don't have much time to read and inspect to fix it

I think it will be much easier to create a new pull request from a separate branch since my changes were simple, is it ok for you?

**Comment by jimczi** [comment]
> I think it will be much easier to create a new pull request from a separate branch since my changes were simple, is it ok for you?

Fine with me, thanks

**Comment by jimczi** [comment]
Superseded by https://github.com/elastic/elasticsearch/pull/42598

**Pull request closed without merging**